scalameter-core/src/main/scala/org/scalameter/execution/invocation/instrumentation/Instrumentation.scala

```diff
@@ -0,0 +1,107 @@
+import scala.collection.convert.decorateAsScala._
+import org.scalameter._
+import org.scalameter.execution.invocation._
+
+
+private[scalameter] object Instrumentation {
+  /** Reads classes from given location and transforms them if they match a predicate.
+   *
+   *  Note that it supports reading from directories and jars.
+   *
+   *  @param in location of classes to read from
+   *  @param p matches if class should be instrumented
+   *  @param instrumentFrom instruments class from a [[java.io.InputStream]]
+   *  @return [[scala.Iterator]] of tuples in a form of (fullClassName, instrumentedClassBytes)
+   */
+  private def transformClasses(in: File)(p: String => Boolean)(instrumentFrom: (String, InputStream) => Array[Byte]): Iterator[(String, Array[Byte])] = {
+    val dirFilter = new FilenameFilter {
+      def accept(file: File, s: String): Boolean = new File(file, s).isDirectory
+    }
+
+    val classFilter = new FilenameFilter {
+      def accept(file: File, s: String): Boolean = s.endsWith(".class")
+    }
+
+    def _transformClasses(in: File, rootPath: String): Iterator[(String, Array[Byte])] = {
+      val subdirs = in.listFiles(dirFilter)
+      val classes = in.listFiles(classFilter)
+      classes.iterator.collect { case f if p(s"$rootPath/${f.getName}".stripPrefix("/").stripSuffix(".class").replace('/', '.')) =>
+        val path = s"$rootPath/${f.getName}".stripPrefix("/")
+        path -> instrumentFrom(path.stripSuffix(".class"), new FileInputStream(f))
+      } ++ subdirs.iterator.flatMap(f => _transformClasses(f, s"$rootPath/${f.getName}".stripPrefix("/")))
+    }
+
+    if (in.isDirectory) {
+      _transformClasses(in, rootPath = "")
+    } else { // if it's not directory try to read it as zip/jar file
+      val zipFile = new ZipFile(in)
+      zipFile.entries().asScala.collect {
+        case entry if p(entry.getName.stripSuffix(".class").replace('/', '.')) =>
+          entry.getName -> instrumentFrom(entry.getName.stripSuffix(".class"), zipFile.getInputStream(entry))
+      }
+    }
+  }
+
+  /** Writes jar with instrumented classes to a given file.
+   *
+   *  Note that if given [[org.scalameter.Context]] does not contain classpath
+   *  [[org.scalameter.utils.ClassPath.default]] is used to get a default classpath.
+   *
+   *  @param ctx [[org.scalameter.Context]] with a classpath key
+   *  @param matcher [[org.scalameter.execution.invocation.InvocationCountMatcher]] to match methods that needs instrumentation
+   *  @param to writes jar with instrumented to a given file
+   *  @return lookup table of instrumented methods
+   */
+  def writeInstrumentedClasses(ctx: Context, matcher: InvocationCountMatcher, to: File): Vector[MethodSignature] = {
+    val classpath = ctx.goe(Key.classpath, utils.ClassPath.default)
+    val lookupTable = Vector.newBuilder[MethodSignature]
+    val instrumented: Iterator[(String, Array[Byte])] = classpath.paths.iterator.flatMap { f =>
+      var currentIndex = 0
```

Comment:
You should either be fully functional and use `zipWithIndex` in the previous line, or you should be imperative, and process all the classes inside a `foreach` loop.

I would advise you to refactor (and rename) the code in the `transformClasses` method to return an `Iterator` over the input classes, instead of pushing the transformation function into that method:

    def filterClasses(in: File)(p: String => Boolean): Iterator[(String, File)] = ...

This gives you several benefits:
- it's easier to debug and understand the code, since the transformation happens in the outer loop instead of in the function object call inside the iterator
- simpler interface for `filterClasses`, which is welcome, since you only use the `filterClasses` once
- better interaction with any mutable state, such as `currentIndex`, which happens as part of the transformation